components/org.wso2.micro.integrator.initializer/src/main/java/org/wso2/micro/integrator/initializer/handler/transaction/TransactionHandlerComponent.java

```diff
@@ -0,0 +1,172 @@
+        }
+    }
+
+    /**
+     * Schedule the transaction writer task to write the current transaction count to the database at a fixed rate.
+     *
+     * @param transactionStore - transactionStore instance.
+     */
+    private void scheduleTransactionWriterTask(TransactionStore transactionStore) {
+        ScheduledExecutorService trnCountWriterTaskScheduler = Executors.newSingleThreadScheduledExecutor();
+        int scheduledPeriod = getTransactionCountScheduledPeriod();
+
+        trnCountWriterTaskScheduler.scheduleAtFixedRate(() -> {
+            try {
+                transactionStore.addTransaction();
+            } catch (TransactionException e) {
+                LOG.warn("Could not persist the transaction count: ", e);
+            }
+        }, scheduledPeriod, scheduledPeriod, TimeUnit.SECONDS);
+    }
+
+    private String generateRandomId() {
+        return UUID.randomUUID().toString();
+    }
+
+    /**
+     * Check whether datasource is configured for transaction counting utility.
+     *
+     * @param dataSourceService dataSourceService instance.
+     * @return - true if datasource defined.
+     * @throws DataSourceException  - when DataSource is not an RDBMS data source.
+     * @throws TransactionException - when no DataSource is available for transaction counting utility.
+     */
+    public DataSource getTransactionDataSource(DataSourceService dataSourceService)
+            throws DataSourceException, TransactionException {
+
+        String dataSourceId = getTransactionDatasourceId();
+        CarbonDataSource dataSource = dataSourceService.getDataSource(dataSourceId);
+        if (dataSource == null) {
+            throw new TransactionException("DataSource " + dataSourceId + " is not available for transaction.");
+        }
+        Object transactionDataSourceObject = dataSource.getDSObject();
+        if (!(transactionDataSourceObject instanceof DataSource)) {
+            throw new DataSourceException("DataSource is not an RDBMS data source.");
+        }
+        return (DataSource) transactionDataSourceObject;
+    }
+
+    /**
+     * Get the dataSource configured for transaction counting utility.
+     *
+     * @return - datasource id.
+     * @throws TransactionException - when a dataSource is not configured for transaction counting utility.
+     */
+    private String getTransactionDatasourceId() throws TransactionException {
+        String dataSourceId = System.getProperty(TRANSACTION_DATA_SOURCE);
+        if (StringUtils.isEmpty(dataSourceId)) {
+            dataSourceId = System.getenv(TRANSACTION_DATA_SOURCE);
+            if (StringUtils.isEmpty(dataSourceId)) {
+                Object dataSourceIdObject = ConfigParser.getParsedConfigs().get(
+                        TRANSACTION_CONFIG_SECTION + "." + TRANSACTION_CONFIG_NAME);
+                if (dataSourceIdObject != null) {
+                    dataSourceId = dataSourceIdObject.toString();
+                } else {
```

Comment:
Shall we default to carbon db if not configured.

Reply:
we cannot default to carbon db since we do not expect everybody to configure it. Not even a majority. We thought of the coordination db, but once again, we cannot expect everybody to have that either.

If the user doesn't have this configuration, it is highly likely that he hasn't created a database tables either. Since this is a conscious thing he has to do, we thought of no defaulting it, WDYT?

Reply:
If this is a conscious thing and we don't provide a default storage , is it correct to enable it by default .

Reply:
Disabling the Transaction count component by default. Hence, if the user wants to enable this feature, the required configurations will give them the notion that a valid data source should be configured priorly during the configuration time.